### PR TITLE
Smithy4s Kinesis Client

### PR DESCRIPTION
**Pull request by etspaceman**
## Changes Introduced

Introduces a Smithy4s Kinesis Client implementation, bundled with logs and localstack support. Important decisions:

- Slf4j was not included as the default logger for this module, as we would like to eventually cross-build this for scalajs/native. Instead, a NoOp logger is the default and users can provide their own implementations as needed. This matches the approach that Http4s is using in their upcoming changes.
- Scoverage was removed as there were issues with running coverage tests with smithy4sGenerate. There is little actual benefit to scoverage, so it feels safe to remove
- sbt-projectmatrix was included to help reason between the scala version differences, as Smithy4s does not offer a 2.12 version of the AWS modules
- Using scalafmt instead of scalafix to organize imports
- sbt-typelevel's MergifyPlugin was included to help manage that file

## Applicable linked issues

#17 #16 

## Checklist (check all that apply)

- [x] This change maintains backwards compatibility
- [x] I have introduced tests for all new features and changes
- [x] I have added documentation covering all new features and changes
- [x] This pull-request is ready for review

### CODE REVIEW

**Comment by etspaceman** [comment]
@kubukoz per our convo, I tried setting up a client using smithy4s, but ran into a few weird bugs with the build. Mostly, I've been dealing with using a subset of scala versions in the smithy project as smithy4s is only publishing Scala 2.13/Scala 3 artifacts, so I am poorly attempting to use sbt-projectmatrix as well. But also, I ran into this when compiling on scala 2.13 versions:

```[error] (smithy4s-client / Compile / smithy4sCodegen) java.lang.ClassCastException: class software.amazon.smithy.rulesengine.traits.ContextParamTrait cannot be cast to class software.amazon.smithy.rulesengine.traits.ContextParamTrait (software.amazon.smithy.rulesengine.traits.ContextParamTrait is in unnamed module of loader java.net.URLClassLoader @17cc56dd; software.amazon.smithy.rulesengine.traits.ContextParamTrait is in unnamed module of loader java.net.URLClassLoader @756a6cba)```

File: project/Kinesis4CatsPlugin.scala [hunk 1]
```diff
@@ -263,6 +276,37 @@ object Kinesis4CatsPluginKeys {
...
+          Defaults.testSettings ++
+          headerSettings(FunctionalTest) ++
+          Seq(
+            parallelExecution := false,
+            javaOptions += "-Dcom.amazonaws.sdk.disableCertChecking=true"
+          )
+      })
+  }
+
   final implicit class Kinesi4catsProjectOps(private val p: Project)
```
Comment (kubukoz) [inline project/Kinesis4CatsPlugin.scala:310]: fyi: typo in class name

**Comment by kubukoz** [comment]
Never seen that before I think... perhaps @baccata will have more insight. I know there was talk of removing the rules engine traits from the model before loading, and in general smithy4s will eventually make it easier to use AWS specs.

https://github.com/disneystreaming/smithy4s/issues/674#issuecomment-1348748220 and that general thread may have more useful context

**Comment by kubukoz** [comment]
`smithy4s-client / Compile / smithy4sCodegen`

I ran that in sbt as `smithy4s-client/compile`, and it worked. CI seems to be failing because the yml generated doesn't match the one checked in, so it's hard for me to see the issue in CI either.

Additionally it'd be great to minimize any issue you encounter with smithy4s, as your build is pretty big and there could be multiple things at play

**Comment by kubukoz** [comment]
if you do manage to hit this again, can you try with a local snapshot of smithy4s from the `series/0.17` branch?

**Comment by etspaceman** [comment]
Thanks for looking into it @kubukoz ! I'll be out of pocket today but I'll see if I can get this PR cleaned up and reproduce the same.

**Comment by Baccata** [comment]
@etspaceman don't bother trying a snapshot, it does fix the bug you're currently encountering but another unrelated bug was introduce which leads the generated code for the `smithy.rules` to be generated twice. I'm in the process of fixing it

**Comment by Baccata** [comment]
Here's a fix to the newly introduced bug I'm talking about : https://github.com/disneystreaming/smithy4s/pull/789

**Comment by etspaceman** [comment]
OK PR is cleaned up with clean reproduce, you should be able to get this via doing `sbt smithy4s-client3/compile`

```
[error] stack trace is suppressed; run last smithy4s-client3 / Compile / smithy4sCodegen for the full output
[error] (smithy4s-client3 / Compile / smithy4sCodegen) java.lang.ClassCastException: class software.amazon.smithy.rulesengine.traits.ContextParamTrait cannot be cast to class software.amazon.smithy.rulesengine.traits.ContextParamTrait (software.amazon.smithy.rulesengine.traits.ContextParamTrait is in unnamed module of loader java.net.URLClassLoader @485438eb; software.amazon.smithy.rulesengine.traits.ContextParamTrait is in unnamed module of loader java.net.URLClassLoader @1f8c81d0)
```

**Comment by kubukoz** [comment]
I can't see it with just `client3`, but I can see it with this sequence:

1. `smithy4s-client/compile`
2. `smithy4s-client3/compile`

Updating to a local snapshot of https://github.com/disneystreaming/smithy4s/pull/789 seems to make it work.

**Comment by etspaceman** [comment]
@kubukoz It does make this particular error go away but then a number of compiler errors come on due to the other bug that @Baccata found. 

It does seem to be working from the PR branch in @Baccata 's proposal! So that's great 😄

File: .scalafix.conf [hunk 2]
```diff
@@ -1,4 +1,4 @@
-rules = [DisableSyntax, NoValInForComprehension, ProcedureSyntax, OrganizeImports]
+rules = [DisableSyntax, NoValInForComprehension, OrganizeImports]
```
Comment (etspaceman) [inline .scalafix.conf:1]: sbt-projectmatrix added the impact of having scalafmt/scalafix operations run for each project inside of the matrix, including scala 3. These updates allow us to run it on scala 3.

File: build.sbt [hunk 3]
```diff
@@ -205,75 +221,131 @@ lazy val `kpl-ciris` = project
...
+      S4S.http4sAws(smithy4sVersion.value),
+      Log4Cats.slf4j,
+      Smithy.rulesEngine % Smithy4s
+    ),
+    Compile / smithy4sAllowedNamespaces := List(
+      "smithy.rules",
+      "com.amazonaws.kinesis"
+    )
+  )
+  .jvmPlatform(last2ScalaVersions)
```
Comment (etspaceman) [inline build.sbt:314]: Brought in sbt-projectmatrix to work through having a subset of scala versions supported here (Smithy4s AWS does not have a 2.12 offering)

File: project/DockerComposePlugin.scala [hunk 4]
```diff
@@ -140,7 +140,7 @@ object DockerComposePlugin extends AutoPlugin {
       configuration: Configuration,
       build: Boolean,
       projects: List[Project]
-  ): Seq[Setting[_]] =
+  ): Seq[Setting[_]] = inConfig(configuration)(
```
Comment (etspaceman) [inline project/DockerComposePlugin.scala:143]: Fixes the integration tests always targeting the functional test file...

File: project/Kinesis4CatsPlugin.scala [hunk 5]
```diff
@@ -186,6 +186,17 @@ object Kinesis4CatsPlugin extends AutoPlugin {
     Test / testOptions ++= {
       List(Tests.Argument(MUnitFramework, "+l"))
     },
+    scalacOptions ++= {
```
Comment (etspaceman) [inline project/Kinesis4CatsPlugin.scala:189]: https://github.com/typelevel/sbt-typelevel/issues/464

File: smithy4s-client-localstack/src/main/scala/kinesis4cats/smithy4s/client/localstack/LocalstackKinesisClient.scala [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+import smithy4s.aws._
+import smithy4s.aws.http4s._
+import smithy4s.aws.kernel.AwsRegion
+
+import kinesis4cats.localstack.LocalstackConfig
+import kinesis4cats.logging.LogEncoder
+import kinesis4cats.smithy4s.client.KinesisClient
+import kinesis4cats.smithy4s.client.middleware._
+
+object LocalstackKinesisClient {
```
Comment (etspaceman) [inline smithy4s-client-localstack/src/main/scala/kinesis4cats/smithy4s/client/localstack/LocalstackKinesisClient.scala:36]: The implementation here is almost identical to the KinesisClient, except it adds an additional middleware to proxy the request, as well as uses some mock credentials.

File: smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/KinesisClient.scala [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  /** Helper class containing required
+    * [[kinesis4cats.logging.LogEncoder LogEncoders]] for the
+    * [[kinesis4cats.smithy4s.client.KinesisClient KinesisClient]]
+    */
+  final class LogEncoders[F[_]](implicit
+      val http4sRequestLogEncoder: LogEncoder[Request[F]],
+      val http4sResponseLogEncoder: LogEncoder[Response[F]]
+  )
+
+  def apply[F[_]](
```
Comment (etspaceman) [inline smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/KinesisClient.scala:43]: Very simple implementation. Basically the provided generated client from Smithy4s with some middleware to add standard logging.

File: smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/middleware/RequestLogger.scala [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+import cats.effect.{Async, Resource}
+import cats.syntax.all._
+import fs2.{Chunk, Pipe, Stream}
+import org.http4s.Response
+import org.http4s.client.Client
+import org.typelevel.log4cats.StructuredLogger
+
+import kinesis4cats.logging.LogContext
+
+object RequestLogger {
```
Comment (etspaceman) [inline smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/middleware/RequestLogger.scala:30]: Mostly copied this from http4s. The http4s implementation records the requests and responses as a part of the message. This one will encode them as a part of the context, per the structured logging desires.

File: smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/middleware/ResponseLogger.scala [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+import cats.effect.{Async, Ref, Resource}
+import cats.syntax.all._
+import fs2.{Chunk, Pipe, Stream}
+import org.http4s.Response
+import org.http4s.client.Client
+import org.typelevel.log4cats.StructuredLogger
+
+import kinesis4cats.logging.LogContext
+
+object ResponseLogger {
```
Comment (etspaceman) [inline smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/middleware/ResponseLogger.scala:29]: [Ditto](https://github.com/etspaceman/kinesis4cats/pull/36/files#r1097968204)

File: smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/middleware/Utils.scala [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package kinesis4cats.smithy4s.client.middleware
+
+import cats.effect.Concurrent
+import cats.syntax.all._
+import org.http4s.{Charset, MediaType, Message}
+import scodec.bits.ByteVector
+
+object Utils {
```
Comment (etspaceman) [inline smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/middleware/Utils.scala:24]: [Ditto](https://github.com/etspaceman/kinesis4cats/pull/36/files#r1097968204)

File: smithy4s-client/src/main/smithy/kinesis.2023-02-01.json [hunk 11]
```diff
@@ -0,0 +1,6884 @@
...
+        }
+      },
+      "com.amazonaws.kinesis#ListStreams": {
+        "type": "operation",
+        "input": {
+          "target": "com.amazonaws.kinesis#ListStreamsInput"
+        },
+        "output": {
+          "target": "com.amazonaws.kinesis#ListStreamsOutput"
+
```
Comment (etspaceman) [inline smithy4s-client/src/main/smithy/kinesis.2023-02-01.json:5072]: TODO: This is supposed to be 1, but it is a workaround due to a bug in kinesis-mock:

https://github.com/etspaceman/kinesis-mock/pull/418/files

Comment (etspaceman) [inline smithy4s-client/src/main/smithy/kinesis.2023-02-01.json:5364]: TODO: This is supposed to be PositiveIntegerOpbect, but it is a workaround with a min of 0 due to a bug in kinesis-mock:

https://github.com/etspaceman/kinesis-mock/pull/418/files

**Comment by etspaceman** [comment]
I've got the implementation here working. Whenever a new version is published for smithy4s, I can merge this

File: smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/KinesisClient.scala [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+      client: Client[F],
+      region: AwsRegion,
+      creds: F[AwsCredentials]
+  )(implicit
+      F: Async[F],
+      LE: LogEncoders[F]
+  ): Resource[F, KinesisClient[F]] = for {
+    logger <- Slf4jLogger.create[F].toResource
```
Comment (Baccata) [inline smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/KinesisClient.scala:51]: My two cents : I'm personally not a fan of slf4j being pulled by default. For starters it prevents you from cross-building to js/native down the road (though I don't think people will care much about js, native could be very interesting). 

Moreover, I just think libraries should never instantiate loggers ever. I think it's fine in this context to standardise against log4cats, but the logger should be injected as a parameter, just like the http client. 

Put differently : you're not choosing the http4s backend for your users, you shouldn't choose the logging backend for them ^^

Comment (etspaceman) [reply smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/KinesisClient.scala:51]: Fair enough. I'll see what I can do about abstracting the logging backend in a different PR.

Comment (etspaceman) [reply smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/KinesisClient.scala:51]: @Baccata Http4s uses the approach of having a default logger of Slf4jLogger, with the ability to override it on their builders. Would that be acceptable here for your concern?

Comment (Baccata) [reply smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/KinesisClient.scala:51]: That's certainly a sensible approach

Comment (kubukoz) [reply smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/KinesisClient.scala:51]: @etspaceman http4s (ember, in particular) does that [on the JVM](https://github.com/http4s/http4s/blob/0e253e30bf983545cb0ebdd6407495b629f4ed1e/ember-server/jvm/src/main/scala/org/http4s/ember/server/EmberServerBuilderPlatform.scala#L25) but you'll see that the [JS/Native implementations](https://github.com/http4s/http4s/blob/0e253e30bf983545cb0ebdd6407495b629f4ed1e/ember-server/js-native/src/main/scala/org/http4s/ember/server/EmberServerBuilderPlatform.scala#L25) use a no-op logger by default. What is more, they're going to [get rid of it on the JVM too](https://github.com/http4s/http4s/pull/6645).

Comment (kubukoz) [reply smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/KinesisClient.scala:51]: I like the idea of using NoOp as a default and having some sort of builder where you can pass a different one.

Comment (etspaceman) [reply smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/KinesisClient.scala]: Alright. I've updated this to use a NoOp implementation by default.

I feel pretty safe leaving the SL4FJ dependency defaults in the java wrapper projects as they are provided by the underlying wrappers to begin with. But for this project, with it being native, it makes sense to have a choice here.

File: smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/KinesisClient.scala [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+    * @param LE
+    *   [[kinesis4cats.smithy4s.client.KinesisClient.LogEncoders KinesisClient.LogEncoders]]
+    * @return
+    *   [[cats.effect.Resource Resource]] of a Kinesis Client.
+    */
+  def apply[F[_]](
+      client: Client[F],
+      region: AwsRegion,
+      loggerF: Async[F] => F[StructuredLogger[F]] = (f: Async[F]) =>
+        f.pure(NoOpLogger[F](f)),
```
Comment (etspaceman) [inline smithy4s-client/src/main/scala/kinesis4cats/smithy4s/client/KinesisClient.scala:86]: Default of NoOpLogger, allows users to configure their own backend for the logging.

File: preprocessors/src/main/scala/preprocessors/KinesisSpecTransformer.scala [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+      .addTrait(RangeTrait.builder().min(java.math.BigDecimal.ZERO).build())
+      .build()
+
+  val traitTransform: BiFunction[Shape, Trait, Trait] =
+    (shape: Shape, `trait`: Trait) =>
+      if (
+        `trait`.toShapeId() == LengthTrait.ID &&
+        shape.toShapeId() == metricsNameListShapeId
+      ) {
+        LengthTrait.builder().min(0).max(7).build()
```
Comment (etspaceman) [inline preprocessors/src/main/scala/preprocessors/KinesisSpecTransformer.scala:44]: Usually this is 1-7. Set to 0-7

File: preprocessors/src/main/scala/preprocessors/KinesisSpecTransformer.scala [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    ShapeId.fromParts("com.amazonaws.kinesis", "MetricsNameList")
+
+  private val putRecordsOutputShapeId =
+    ShapeId.fromParts("com.amazonaws.kinesis", "PutRecordsOutput")
+
+  private val nonNegativeIntegerObjectShape =
+    IntegerShape
+      .builder()
+      .id(
+        ShapeId.fromParts("com.amazonaws.kinesis", "NonNegativeIntegerObject")
```
Comment (etspaceman) [inline preprocessors/src/main/scala/preprocessors/KinesisSpecTransformer.scala:33]: Counterpart to PositiveIntegerObject

File: smithy4s-client-transformers/src/main/scala/kinesis4cats/smithy4s/client/KinesisSpecTransformer.scala [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+      .addTrait(RangeTrait.builder().min(java.math.BigDecimal.ZERO).build())
+      .build()
+
+  val traitTransform: BiFunction[Shape, Trait, Trait] =
+    (shape: Shape, `trait`: Trait) =>
+      if (
+        `trait`.toShapeId() == LengthTrait.ID &&
+        shape.toShapeId() == metricsNameListShapeId
+      ) {
+        LengthTrait.builder().min(0).max(7).build()
```
Comment (etspaceman) [inline smithy4s-client-transformers/src/main/scala/kinesis4cats/smithy4s/client/KinesisSpecTransformer.scala:67]: Empty responses are possible, this seems to be a bug w/ the smithy file

File: smithy4s-client-transformers/src/main/scala/kinesis4cats/smithy4s/client/KinesisSpecTransformer.scala [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+      } else if (
+        shape
+          .toShapeId() == listStreamConsumersOutputNextTokenShapeId && `trait`
+          .toShapeId() == DocumentationTrait.ID
+      ) {
+        new DocumentationTrait(
+          `trait`
+            .asInstanceOf[DocumentationTrait] // scalafix:ok
+            .getValue()
+            .replace("ListStreamConsumersInput$NextToken", "NextToken")
```
Comment (etspaceman) [inline smithy4s-client-transformers/src/main/scala/kinesis4cats/smithy4s/client/KinesisSpecTransformer.scala:77]: The dollar sign is not valid for scaladoc, so we replace it here.

File: smithy4s-client-transformers/src/main/scala/kinesis4cats/smithy4s/client/KinesisSpecTransformer.scala [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+      } else if (
+        shape
+          .toShapeId() == listShardsOutputNextTokenShapeId && `trait`
+          .toShapeId() == DocumentationTrait.ID
+      ) {
+        new DocumentationTrait(
+          `trait`
+            .asInstanceOf[DocumentationTrait] // scalafix:ok
+            .getValue()
+            .replace("ListShardsInput$NextToken", "NextToken")
```
Comment (etspaceman) [inline smithy4s-client-transformers/src/main/scala/kinesis4cats/smithy4s/client/KinesisSpecTransformer.scala:88]: [https://github.com/etspaceman/kinesis4cats/pull/36/files#r1103718254](Ditto)

File: smithy4s-client-transformers/src/main/scala/kinesis4cats/smithy4s/client/KinesisSpecTransformer.scala [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+        shape
+          .getAllMembers()
+          .entrySet()
+          .stream()
+          .map[MemberShape] { x =>
+            val (memberName, member) = (x.getKey(), x.getValue())
+            if (memberName == "FailedRecordCount")
+              MemberShape
+                .builder()
+                .target(nonNegativeIntegerObjectShape.getId())
```
Comment (etspaceman) [inline smithy4s-client-transformers/src/main/scala/kinesis4cats/smithy4s/client/KinesisSpecTransformer.scala:104]: FailedRecordCount can be 0. This is a bug with smithy

**Pull request merged**